bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/XmlSequence.java

```diff
@@ -671,9 +699,8 @@ public boolean equals(Object o, Set<ValuePair> visitedValues) {
         if (o instanceof XmlSequence rhsXMLSequence) {
             return isXMLSequenceChildrenEqual(this.getChildrenList(), rhsXMLSequence.getChildrenList());
         }
-        if (o instanceof XmlItem) {
-            return this.getChildrenList().size() == 1 &&
-                    isEqual(this.getChildrenList().get(0), o);
+        if (this.isSingleton() && (o instanceof XmlItem || o instanceof XmlText)) {
```

Comment:
Still won't work for things like

```ballerina
xml x1 = xml:concat(xml `<!--foo-->`, xml ``);
xml:Comment x2 = xml `<!--foo-->`;
boolean equal = x1 == x2;
```

Reply:
Fixed with dd8e19f2a0e5255dbe863e931bffe66bf0190024